tests/unit/select_connection_ioloop_tests.py

```diff
@@ -81,7 +89,54 @@ def start_test(self):
         self.start()
 
     def on_timer(self, val):
-        self.assertEquals(val, self.timer_stack.pop())
+        self.assertEqual(val, self.timer_stack.pop())
+        if not self.timer_stack:
+            self.ioloop.stop()
+
+    def test_normal(self):
+        self.start_test()
+
+    def test_timer_for_suicide(self):
+        self.timer_stack = list()
+        handle_holder = []
+        self.handle = self.ioloop.add_timeout(
+            0.1, partial(self.on_timer_suicide,handle_holder))
+        handle_holder.append(self.handle)
+        self.start()
+
+    def on_timer_suicide(self, handle_holder):
+        self.assertEqual(self.handle, handle_holder.pop())
+        # This removal here should not raise exception by itself nor
+        # in the caller SelectPoller.process_timeouts().
+        self.ioloop.remove_timeout( self.handle )
+        self.ioloop.stop()
+
+    # This patch delays the execution of the handlers so that they get fired
+    # in the same invocation of SelectPoller.process_timeouts().
+    @patch('pika.adapters.select_connection.SelectPoller.is_ready_to_fire',
```

Comment:
Please get rid of `SelectPoller.is_ready_to_fire` from SelectPoller class, this patch, and the "This patch delays the execution..." comment above. They are not needed to test this functionality. The following should suffice to ensure that the timers will be ready in the same invocation of `SelectPoller.process_timeouts()`:

0. Initialize `self.deleted_another_timer = False`
1. Create timer 1: 0.01 seconds: callback is closure (function in function) `on_timer_that_deletes_another_ready_timer`. This callback should delete timer 2 and leave a breadcrumb in self (e.g., `self.deleted_another_timer = True`)
2. Create timer 2: 0.02 seconds: callback is closure `on_deleted_timer`. This callback should call `self.fail('deleted timer's callback called')`
3. Create timer 3: 0.03 seconds: callback is closure `on_conclude_another_timer_deletion`. This callback checks `self.assertTrue(self.deleted_another_timer)`, checks `self.assertNotIn(timer2_id, self.ioloop._timeouts)`, and calls `self.ioloop.stop()`.